src/DmaRequestCore.bsv

```diff
@@ -22,78 +24,80 @@ module mkChunkComputer(ChunkCompute ifc);
 
     Reg#(DmaMemAddr) newChunkPtrReg <- mkReg(0);
     Reg#(DmaMemAddr) totalLenRemainReg <- mkReg(0);
-    Reg#(Bool) isSplittigReg <- mkReg(False);
+    Reg#(Bool) isSplittingReg <- mkReg(False);
 
     function Bool hasBoundary(DmaRequestFrame request);
-        let highIdx = (request.startAddr + request.length) >> BUS_BOUNDARY_WIDTH;
-        let lowIdx = request.startAddr >> BUS_BOUNDARY_WIDTH;
+        let highIdx = (request.startAddr + request.length - 1) >> valueOf(BUS_BOUNDARY_WIDTH);
+        let lowIdx = request.startAddr >> valueOf(BUS_BOUNDARY_WIDTH);
         return (highIdx > lowIdx);
     endfunction
 
     function DmaMemAddr getOffset(DmaRequestFrame request);
-        DmaMemAddr offset = zeroExtend(fromInteger(valueOf(BUS_BOUNDARY)) - pack(request.startAddr[BUS_BOUNDARY_WIDTH-1:0]));
-        return offset;
+        // 4096 - startAddr % 4096
+        Bit#(BUS_BOUNDARY_WIDTH) remainder = truncate(request.startAddr);
+        Bit#(BUS_BOUNDARY_WIDTH) offset = fromInteger(valueOf(BUS_BOUNDARY)-1) - zeroExtend(remainder) + 1;
+        return zeroExtend(offset);
     endfunction
 
-    rule getfirstChunkLen if(inputFifo.notEmpty && splitFifo.notFull);
+    rule getfirstChunkLen;
         let request = inputFifo.first;
         inputFifo.deq;
-        let offset = getOffSet(request);
+        let offset = getOffset(request);
         // firstChunkLen = offset % PCIE_TLP_BYTES
-        DmaMemAddr firstLen = zeroExtend(offset[valueOf(PCIE_TLP_BYTES_WIDTH)-1:0]);
-        ChunkRequestFrame splitRequest = {
+        Bit#(PCIE_TLP_BYTES_WIDTH) offsetModTlpBytes = truncate(offset);
+        DmaMemAddr firstLen = zeroExtend(offsetModTlpBytes);
+        splitFifo.enq(ChunkRequestFrame {
             dmaRequest: request,
-            firstChunkLen: hasBoundary(request) ? tagged Valid firstLen : tagged Invalid; 
-        }
-        splitFifo.enq(splitRequest);
-    endrule
+            firstChunkLen: hasBoundary(request) ? tagged Valid firstLen : tagged Invalid
```

Comment:
use a `Maybe` suffix to indicate this variable holds a maybe value. so, rename it to `firstChunkLenMaybe`

Reply:
OK.